stage #2.5/module #3 Spring Core/#5.1. Annotation-based configuration/content.md

```diff
@@ -0,0 +1,470 @@
+
+Even typed `Map` instances can be autowired as long as the expected key type is String. The map values contain all beans
+of the expected type, and the keys contain the corresponding bean names, as the following example shows:
+
+```java
+public class MusicRatingService {
+
+    private Map<String, MusicMagazine> musicMagazines;
+
+    @Autowired
+    public void setMusicMagazines(Map<String, MusicMagazine> musicMagazines) {
+        this.musicMagazines = musicMagazines;
+    }
+
+    // ...
+}
+```
+
+By default, autowiring fails when no matching candidate beans are available for a given injection point. In the case of
+a declared array, collection, or map, at least one matching element is expected.
+
+The default behavior is to treat annotated methods and fields as indicating required dependencies. You can change this
+behavior, enabling the framework to skip a non-satisfiable injection point through marking it as non-required (i.e., by
+setting the required attribute in `@Autowired` to false):
+
+```
+@Autowired(required = false)
+```
+
+A non-required method will not be called at all if its dependency (or one of its dependencies, in case of multiple
+arguments) is not available. A non-required field will not get populated at all in such cases, leaving its default value
+in place.
+
+Alternatively, you can express the non-required nature of a particular dependency through Java 8's `java.util.Optional`,
+as the following example shows:
+
+```java
+public class MusicRatingService {
+
+    private Optional<MusicRatesFinder> musicRatesFinder;
+
+    @Autowired
+    public void setMusicRatesFinder(Optional<MusicRatesFinder> musicRatesFinder) {
+        this.musicRatesFinder = musicRatesFinder;
+    }
+
+    // ...
+}
+```
+
+As of Spring Framework 5.0, you can also use a `@Nullable annotation` (of any kind in any package - for example,
+`javax.annotation.Nullable` from JSR-305)
+
+```java
+public class MusicRatingService {
+
+    private MusicRatesFinder musicRatesFinder;
+
+    @Autowired
+    public void setMusicRatesFinder(@Nullable MusicRatesFinder musicRatesFinder) {
+        this.musicRatesFinder = musicRatesFinder;
+    }
+
+    // ...
+}
+```
+
+You can also use `@Autowired` for interfaces that are well-known resolvable dependencies: `BeanFactory`,
+`ApplicationContext`, `Environment`, `ResourceLoader`, `ApplicationEventPublisher`, and `MessageSource`. These
+interfaces and their extended interfaces, such as `ConfigurableApplicationContext` or `ResourcePatternResolver`, are
+automatically resolved, with no special setup necessary.
+
+> The `@Autowired`, `@Inject`, `@Value`, and `@Resource` annotations are handled by Spring `BeanPostProcessor`
+> implementations. This means that you cannot apply these annotations within your own `BeanPostProcessor` or
+> `BeanFactoryPostProcessor` types (if any). These types must be 'wired up' explicitly by using XML or a Spring
+> `@Bean` method.
+
+## Annotation-based Autowiring with Qualifiers
```

Comment:
Add somewhere the reason to use @Qualifier: If more than one bean of the same type is available in the container, the framework will throw NoUniqueBeanDefinitionException.  This is because Spring doesn't know which bean to inject.  To avoid this problem, there are several solutions; the @Qualifier annotation is one of them.
I think "more control over the bean candidates selection process" is not clear reason

Reply:
Thanks a lot for the already prepared reason, will use it)